File: patch_units.pl
use strict;use warnings;
use File::Slurp qw(read_file write_file);

sub patch_file {
  my ($path) = @_;
  my $s = read_file($path, binmode => ':utf8');

  # 1) estimateCostUnits: add FORCE_NEXT_UNITS check
  $s =~ s/function\s+estimateCostUnits\(text\)\{\n\s*const\s+s\s*=\s*String\(text\s*\|\|\s*''\s*\);/function estimateCostUnits(text){\n  if (FORCE_NEXT_UNITS) return FORCE_NEXT_UNITS;\n  const s = String(text || '');/m
    or die "patch estimateCostUnits header failed in $path";

  # 2) insert helper block right after estimateCostUnits function definition end (first occurrence)
  my $helper = <<'JS';

// ===== force unit costing for document buttons (SBAR/SOAP/転倒報告) =====
let FORCE_NEXT_UNITS = null;
function setForceNextUnits(units){
  FORCE_NEXT_UNITS = Number(units) || null;
  updateSendButtonCost();
}
function clearForceNextUnits(){
  FORCE_NEXT_UNITS = null;
  updateSendButtonCost();
}
function isUnlockedNow(){
  try{ return !!ls.get("aiga_unlocked_"+VARIANT, false); }catch(e){ return false; }
}
function updateSendButtonCost(){
  try{
    const btn = $("btnSend");
    if (!btn) return;
    if (!isUnlockedNow()){
      btn.textContent = "送信";
      return;
    }
    const txt = String($("userInput")?.value || "").trim();
    const u = txt ? estimateCostUnits(txt) : 1;
    btn.textContent = `送信（-${u}unit）`;
  }catch(e){}
}

JS

  $s =~ s/(function\s+estimateCostUnits\(text\)\{[\s\S]*?\n\})/$1$helper/ or die "insert helper block failed in $path";

  # 3) setTrialInfo: call updateSendButtonCost
  $s =~ s/(function\s+setTrialInfo\(\)\{[\s\S]*?\n\s*\$\("trialInfo"\)\.textContent\s*=\s*[^;]+;)/$1\n  try{ updateSendButtonCost(); }catch(e){};/
    or die "patch setTrialInfo failed in $path";

  # 4) clear helpers: clearForceNextUnits()
  for my $fn (qw(clearSbar clearSoap clearFall)){
    $s =~ s/(function\s+$fn\(\)\{[\s\S]*?)(\n\})/$1\n  try{ clearForceNextUnits(); }catch(e){}\n  try{ updateSendButtonCost(); }catch(e){}$2/
      or die "patch $fn failed in $path";
  }

  # 5) build buttons: setForceNextUnits(LONG_UNIT_COST) after putting text into userInput
  $s =~ s/(\$\("userInput"\)\.value\s*=\s*sbarTxt;)/$1\n        setForceNextUnits(LONG_UNIT_COST);/ or die "patch sbar handler failed in $path";

  # SOAP: input.value = txt;
  $s =~ s/(input\.value\s*=\s*txt;)/$1\n          setForceNextUnits(LONG_UNIT_COST);/ or die "patch soap handler failed in $path";

  # Fall: input.value = txt;
  # There are two occurrences (fall & maybe others). We only want the one in fall build handler; safest: replace the first one after "const txt = buildFallText".
  $s =~ s/(const\s+txt\s*=\s*buildFallText\([^\)]+\);[\s\S]*?\n\s*if\s*\(input\)\{\n\s*input\.value\s*=\s*txt;)/$1\n      setForceNextUnits(LONG_UNIT_COST);/ or die "patch fall handler failed in $path";

  # 6) init: add input listener + initial updateSendButtonCost after binding send
  $s =~ s/(\$\("btnSend"\)\.addEventListener\("click",\s*send\);)/$1\n    $("userInput").addEventListener("input", ()=>{\n      try{ if (FORCE_NEXT_UNITS && !String($("userInput")?.value||"").trim()) clearForceNextUnits(); }catch(e){}\n      try{ updateSendButtonCost(); }catch(e){}\n    });\n    try{ updateSendButtonCost(); }catch(e){};/
    or die "patch init input listener failed in $path";

  # 7) send(): clearForceNextUnits on success, and update button after clearing
  # Insert right after applyChatResult(...);
  $s =~ s/(applyChatResult\(result\);)/$1\n\n    // consume forced long-unit after successful send\n    try{ clearForceNextUnits(); }catch(e){};/
    or die "patch send success clear failed in $path";

  # After userInput.value=""; add updateSendButtonCost
  $s =~ s/(\$\("userInput"\)\.value\s*=\s*"";)/$1\n    try{ updateSendButtonCost(); }catch(e){};/
    or die "patch send clear button update failed in $path";

  # 8) retryLastSend(): on success clear force and update button
  $s =~ s/(applyChatResult\(result\);\n\s*\$\("userInput"\)\.value\s*=\s*"";)/applyChatResult(result);\n\n      try{ clearForceNextUnits(); }catch(e){}\n      $("userInput").value = "";\n      try{ updateSendButtonCost(); }catch(e){};/
    or die "patch retryLastSend failed in $path";

  write_file($path, { binmode => ':utf8' }, $s);
}

patch_file($ARGV[0]);

File: patch_units.js
const fs = require('fs');
const path = require('path');

function patchIndex(file){
  let s = fs.readFileSync(file, 'utf8');

  // 1) Add FORCE_NEXT_UNITS check in estimateCostUnits
  if (!s.includes('if (FORCE_NEXT_UNITS) return FORCE_NEXT_UNITS;') && s.includes('function estimateCostUnits(text){')){
    s = s.replace('function estimateCostUnits(text){\n', 'function estimateCostUnits(text){\n  if (FORCE_NEXT_UNITS) return FORCE_NEXT_UNITS;\n');
  }

  // 2) Insert helper block after estimateCostUnits definition
  const helperBlock = `\n\n// ===== force unit costing for document buttons (SBAR/SOAP/転倒報告/申し送り文) =====\nlet FORCE_NEXT_UNITS = null;\nfunction setForceNextUnits(units){\n  FORCE_NEXT_UNITS = Number(units) || null;\n  updateSendButtonCost();\n}\nfunction clearForceNextUnits(){\n  FORCE_NEXT_UNITS = null;\n  updateSendButtonCost();\n}\nfunction isUnlockedNow(){\n  try{ return !!ls.get(\"aiga_unlocked_\"+VARIANT, false); }catch(e){ return false; }\n}\nfunction updateSendButtonCost(){\n  try{\n    const btn = $(\"btnSend\");\n    if (!btn) return;\n    if (!isUnlockedNow()){\n      btn.textContent = \"送信\";\n      return;\n    }\n    const txt = String($(\"userInput\")?.value || \"\").trim();\n    const u = txt ? estimateCostUnits(txt) : 1;\n    btn.textContent = ` + "`" + `送信（-${u}unit）` + "`" + `;\n  }catch(e){}\n}\n`;

  if (!s.includes('function setForceNextUnits') && s.includes('function estimateCostUnits(text){')){
    // insert right after the first closing brace of estimateCostUnits
    s = s.replace(/(function\s+estimateCostUnits\(text\)\{[\s\S]*?return\s+isLong\s*\?\s*LONG_UNIT_COST\s*:\s*1;\n\})/, `$1${helperBlock}`);
  }

  // 3) Ensure setTrialInfo updates send button label
  if (s.includes('function setTrialInfo(){') && !s.includes('updateSendButtonCost();', s.indexOf('function setTrialInfo(){'))){
    s = s.replace(/(function\s+setTrialInfo\(\)\{[\s\S]*?\$\("trialInfo"\)\.textContent\s*=\s*[^;]+;)/, `$1\n  try{ updateSendButtonCost(); }catch(e){}`);
  }

  // 4) Clear force when panels cleared
  const clearFns = ['clearSbar','clearSoap','clearFall'];
  for (const fn of clearFns){
    const re = new RegExp(`function\\s+${fn}\\(\\)\\{([\\s\\S]*?)\\n\\}`, 'm');
    const m = s.match(re);
    if (m && !m[1].includes('clearForceNextUnits')){
      const replaced = `function ${fn}(){${m[1]}\n  try{ clearForceNextUnits(); }catch(e){}\n  try{ updateSendButtonCost(); }catch(e){}\n}`;
      s = s.replace(re, replaced);
    }
  }

  // 5) Force 5 unit when doc buttons fill input
  if (!s.includes('setForceNextUnits(LONG_UNIT_COST);')){
    s = s.replace(/\$\("userInput"\)\.value\s*=\s*sbarTxt;\n/g, `$("userInput").value = sbarTxt;\n      setForceNextUnits(LONG_UNIT_COST);\n`);
    s = s.replace(/input\.value\s*=\s*txt;\n/g, `input.value = txt;\n          setForceNextUnits(LONG_UNIT_COST);\n`);
  }

  // 6) Add input listener + initial update
  if (!s.includes('addEventListener("input"') && s.includes('$("userInput").addEventListener("keydown"')){
    s = s.replace(/\$\("userInput"\)\.addEventListener\("keydown",\s*\(e\)=>\{\n\s*if\s*\(\(e\.ctrlKey\s*\|\|\s*e\.metaKey\)\s*&&\s*e\.key\s*===\s*"Enter"\)\s*send\(\);\n\s*\}\);/,
`$("userInput").addEventListener("keydown", (e)=>{\n      if ((e.ctrlKey || e.metaKey) && e.key === "Enter") send();\n    });\n    $("userInput").addEventListener("input", ()=>{\n      try{ if (FORCE_NEXT_UNITS && !$("userInput").value.trim()) clearForceNextUnits(); }catch(e){}\n      try{ updateSendButtonCost(); }catch(e){}\n    });\n    updateSendButtonCost();`);
  }

  // 7) Clear force on successful send/retry and refresh button
  // send path
  s = s.replace(/await\s+applyChatResult\(out,\s*\{\s*text,\s*scene,\s*persona,\s*category,\s*level,\s*userItem,\s*aiItem,\s*meta,\s*costUnits\s*\}\);/,
`await applyChatResult(out, { text, scene, persona, category, level, userItem, aiItem, meta, costUnits });\n      try{ clearForceNextUnits(); }catch(e){}\n      try{ updateSendButtonCost(); }catch(e){}`);

  // retry path
  s = s.replace(/await\s+applyChatResult\(out,\s*p\);/g, `await applyChatResult(out, p);\n    try{ clearForceNextUnits(); }catch(e){}\n    try{ updateSendButtonCost(); }catch(e){}`);

  // when input cleared
  s = s.replace(/\$\("userInput"\)\.value\s*=\s*"";\n/g, `$("userInput").value = "";\n      try{ updateSendButtonCost(); }catch(e){}\n`);

  fs.writeFileSync(file, s, 'utf8');
}

const targets = [
  '/mnt/data/work/app/ssw/index.html',
  '/mnt/data/work/app/trainee/index.html'
];

for (const f of targets){
  if (fs.existsSync(f)){
    patchIndex(f);
    console.log('patched', f);
  } else {
    console.error('missing', f);
  }
}
